rust/load/src/bit_difference.rs

```diff
@@ -0,0 +1,459 @@
+//! cluster (this is why being quick to seek is important).
+//!
+//! Internally, guacamole provides primitives that make it easy to manage the set of seeds to get a
+//! variety of data sets out of the synthetic data.
+
+use chromadb::v2::collection::{CollectionEntries, GetOptions, QueryOptions};
+use chromadb::v2::ChromaClient;
+use guacamole::combinators::*;
+use guacamole::{FromGuacamole, Guacamole, Zipf};
+use siphasher::sip::SipHasher24;
+use tracing::Instrument;
+
+use crate::words::WORDS;
+use crate::{DataSet, GetQuery, KeySelector, QueryQuery, Skew, UpsertQuery};
+
+const EMBEDDING_BYTES: usize = 128;
+const EMBEDDING_SIZE: usize = 8 * EMBEDDING_BYTES;
+
+/// This magic constant for hashing gives a consistent way to compute the document ID from the
+/// document content when paired with siphasher.
+pub const MAGIC_CONSTANT_FOR_HASHING: [u8; 16] = [
+    0x63, 0x68, 0x72, 0x6f, 0x6d, 0x61, 0x20, 0x62, 0x65, 0x6e, 0x63, 0x68, 0x6d, 0x61, 0x72, 0x6b,
+];
+
+fn embedding(embedding: [u8; EMBEDDING_BYTES]) -> Vec<f32> {
+    let mut result = vec![];
+    for byte in embedding.into_iter() {
+        for j in 0..8 {
+            if byte & (1 << j) != 0 {
+                result.push(1.0);
+            } else {
+                result.push(0.0);
+            }
+        }
+    }
+    result
+}
+
+/// Options for generating synthetic data.
+#[derive(Clone, Debug)]
+pub struct EmbeddingOptions {
+    /// The number of clusters to generate.
+    pub num_clusters: usize,
+    /// The seed for the random number generator for the clusters.  Given the same seed, the
+    /// clusters will overlap.  For example, if you generate 10 clusters with the same seed and
+    /// then update the num_clusters to 100, you will generate the same 10 clusters and 90 new
+    /// clusters.
+    pub seed_clusters: usize,
+    /// Clustering options.
+    pub clustering: ClusterOptions,
+}
+
+/// Options for generating a single cluster.
+#[derive(Clone, Debug)]
+pub struct ClusterOptions {
+    /// The maximum number of adjacent documents.
+    pub max_adjacent: u32,
+    /// The theta of the zipf distribution for the number of adjacent documents.
+    pub adjacent_theta: f64,
+}
+
+/// A document is a string of content.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Document {
+    /// The content of the document.
+    pub content: String,
+}
+
+impl Document {
+    /// The ID of the document.  This ID is a deterministic function of document content.
+    pub fn id(&self) -> String {
+        let hasher = SipHasher24::new_with_key(&MAGIC_CONSTANT_FOR_HASHING);
+        let h = hasher.hash(self.content.as_bytes());
+        format!("doc:{}", h)
+    }
+
+    /// The embedding of the document.  This embedding is a deterministic function of document
+    /// content.
+    pub fn embedding(&self) -> Vec<f32> {
+        let mut result = vec![];
+        let words = self.content.split_whitespace().collect::<Vec<_>>();
+        for word in WORDS.iter() {
```

Comment:
what happens if the wordlist changes? or will created collections never be re-used across chroma load invocations?

Reply:
The words are set in the binary.  They will "never" change.

Reply:
right, they won't change for the same binary
I was asking if there would be a problem if we decided to change the wordlist in a new version or made it longer to simulate a larger embedding space

Reply:
I'd make that a new data set.  Any limitations there?

Reply:
no, that makes sense 👍